packages/compiler/core/parser.ts

```diff
@@ -1070,6 +1142,618 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
+      | ProjectionMemberExpressionNode
+      | IdentifierNode;
+  }
+
+  function parseProjectionMemberExpressionRest(
+    expr: ProjectionExpression,
+    pos: number
+  ): ProjectionExpression {
+    while (token() !== Token.EndOfFile) {
+      if (parseOptional(Token.Dot)) {
+        expr = {
+          kind: SyntaxKind.ProjectionMemberExpression,
+          base: expr,
+          id: parseIdentifier(),
+          selector: ".",
+          ...finishNode(pos),
+        };
+      } else if (parseOptional(Token.ColonColon)) {
+        expr = {
+          kind: SyntaxKind.ProjectionMemberExpression,
+          base: expr,
+          id: parseIdentifier(),
+          selector: "::",
+          ...finishNode(pos),
+        };
+      } else {
+        break;
+      }
+    }
+
+    return expr;
+  }
+
+  function parseProjectionPrimaryExpression(): ProjectionExpression {
+    switch (token()) {
+      case Token.IfKeyword:
+        return parseProjectionIfExpression();
+      case Token.NumericLiteral:
+        return parseNumericLiteral();
+      case Token.StringLiteral:
+        return parseStringLiteral();
+      case Token.TrueKeyword:
+      case Token.FalseKeyword:
+        return parseBooleanLiteral();
+      case Token.OpenBracket:
+        return parseProjectionTupleExpression();
+      case Token.OpenBrace:
+        return parseProjectionModelExpression();
+      case Token.OpenParen:
+        return parseProjectionLambdaOrParenthesizedExpression();
+      case Token.VoidKeyword:
+        return parseVoidKeyword();
+      case Token.NeverKeyword:
+        return parseNeverKeyword();
+      default:
+        return parseIdentifier("expression");
+    }
+  }
+
+  function parseProjectionLambdaOrParenthesizedExpression(): ProjectionExpression {
+    const pos = tokenPos();
+    const exprs = parseList(ListKind.ProjectionExpression, parseProjectionExpression);
+    if (token() === Token.EqualsGreaterThan) {
+      // unpack the exprs (which should be just identifiers) into a param list
+      const params: ProjectionLambdaParameterDeclarationNode[] = [];
+      for (const expr of exprs) {
+        if (expr.kind === SyntaxKind.Identifier) {
+          params.push({
+            kind: SyntaxKind.ProjectionLambdaParameterDeclaration,
+            id: expr,
+            pos: expr.pos,
+            end: expr.end,
+          });
+        } else {
+          error({ code: "token-expected", messageId: "identifier", target: expr });
+        }
+      }
+
+      return parseProjectionLambdaExpressionRest(pos, params);
+    } else {
+      if (exprs.length === 0) {
+        error({
+          code: "token-expected",
+          messageId: "expression",
+        });
+      }
+      // verify we only have one entry
+      for (let i = 1; i < exprs.length; i++) {
+        // TODO: this error is nonsense, want to emit an error on the first comma
+        // but unsure how to get its exact position after we've parsed.
+        error({
```

Comment:
todo

Reply:
I've removed the todo since the error experience is ok (it's nonsense only in a pedantic sense).